### PR TITLE
Add applying filter/validating filter logic to live metrics filtering

### PR DESCRIPTION
**Pull request by lzchen**
Part 2 of live metrics filtering. Adds applying of filter logic and validation.

TODO

1. Derive exception metrics from span events
2. Filtering for documents on right hand pane

### CODE REVIEW

File: sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py [hunk 1]
```diff
@@ -258,15 +242,217 @@ def _check_metric_filters(metric_infos: List[DerivedMetricInfo], data: _Telemetr
...
+        return True
+    name = name.lower()
+    if telemetry_type == TelemetryType.DEPENDENCY:
+        if name not in _DEPENDENCY_DATA_FIELD_NAMES:
+            return False
+    elif telemetry_type == TelemetryType.REQUEST:
+        if name not in _REQUEST_DATA_FIELD_NAMES:
+            return False
+    elif telemetry_type == TelemetryType.EXCEPTION:
+        if name not in ("exception.message", "exception.stacktrace"):
```
Comment (harsimar) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py:383]: The UI has these names come through as Exception.Message & Exception.StackTrace
just making sure the casing doesn't cause any issues when it comes to filtering

File: sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py [hunk 2]
```diff
@@ -258,15 +242,217 @@ def _check_metric_filters(metric_infos: List[DerivedMetricInfo], data: _Telemetr
...
+            return False
+    elif telemetry_type == TelemetryType.TRACE:
+        if name != "message":
```
Comment (harsimar) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py:386]: same here, casing from the UI is "Message"

File: sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py [hunk 3]
```diff
@@ -425,3 +456,186 @@ def test_get_metrics_from_projections(self, projection_map_mock):
                 ("test-id4", 4.0),
             ],
         )
+
+    def test_validate_info_invalid_telemetry_type(self):
+        valid_metric_info = mock.Mock(telemetry_type="Request", projection=None, filter_groups=[])
```
Comment (harsimar) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py:461]: naming a little confusing here - this is testing for invalid scenarios, but the derivedMetricInfo is named valid_

File: sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py [hunk 4]
```diff
@@ -425,3 +456,186 @@ def test_get_metrics_from_projections(self, projection_map_mock):
...
+        time_stamp = "100.0:0:0"
+        expected_ms = 8640000000  # 100 days
+        self.assertEqual(_filter_time_stamp_to_ms(time_stamp), expected_ms)
+
+    def test_invalid_time_stamp(self):
+        time_stamp = "invalid_format"
+        self.assertIsNone(_filter_time_stamp_to_ms(time_stamp))
+
+    def test_time_stamp_missing_seconds(self):
+        time_stamp = "1.0:0:"
```
Comment (harsimar) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py:522]: Have you seen this case happen?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py [hunk 5]
```diff
@@ -425,3 +456,186 @@ def test_get_metrics_from_projections(self, projection_map_mock):
...
+        self.assertIsNone(_filter_time_stamp_to_ms(time_stamp))
+
+    def test_time_stamp_empty_string(self):
```
Comment (harsimar) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py:525]: Is this possible?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py [hunk 6]
```diff
@@ -425,3 +456,186 @@ def test_get_metrics_from_projections(self, projection_map_mock):
...
+        valid_metric_info = mock.Mock(telemetry_type="Request", projection=None, filter_groups=[])
+        valid_metric_info.telemetry_type = "Request"
+        valid_metric_info.projection = "CustomMetrics.MyMetric"
+        self.assertFalse(_validate_derived_metric_info(valid_metric_info))
+
+    @mock.patch("azure.monitor.opentelemetry.exporter._quickpulse._utils._validate_filter_predicate_and_comparand")
+    @mock.patch("azure.monitor.opentelemetry.exporter._quickpulse._utils._validate_filter_field_name")
+    def test_validate_info_valid_filters(self, field_mock, pred_mock):
+        valid_metric_info = mock.Mock(telemetry_type="Request", projection=None, filter_groups=[])
+        filter = mock.Mock()
```
Comment (harsimar) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py:475]: I'm not too familiar with the python mocking framework - where is the filter mock actually defined? Ie how do I know what it is filtering for

File: sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py [hunk 7]
```diff
@@ -425,3 +456,186 @@ def test_get_metrics_from_projections(self, projection_map_mock):
...
+        data_mock = mock.Mock()
+        check_mock.return_value = True
+        result = _check_filters([filter_info], data_mock)
+        self.assertTrue(result)
+        check_mock.assert_called_once_with(filter_info, data_mock)
+
+    @mock.patch("azure.monitor.opentelemetry.exporter._quickpulse._utils._check_custom_dim_field_filter")
+    def test_check_custom_dim_filters(self, check_mock):
+        filter_info = mock.Mock()
+        filter_info.field_name = "CustomDimensions."
```
Comment (harsimar) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py:549]: Wouldn't this just give you an empty custom dimension field name?
When a customer enters a custom dim in the UI, it usually appears like this in the response: CustomDimensions.<fieldEntered>

**Comment by harsimar** [comment]
A general comment about casing of field names - 
I noticed in your tests & in parts of your filtering implementation, the case of the fields are all lowercase. Just a reminder that inside the response body of ping/post , the field names are not lowercase there. When reading the config you may have converted things to lowercase for ease in filtering but I didn't see it (feel free to correct if I missed it!). Just make sure that the casing of whatever the original config is doesn't impact the ability for it to find the right column to validate/filter.

**Comment by harsimar** [comment]
Another question - I noticed your validation code returns true/false instead of tracking errors to eventually put inside a CollectionConfigurationErrors on the next post request: https://github.com/aep-health-and-standards/Telemetry-Collection-Spec/blob/main/ApplicationInsights/live-metrics-filtering.md#what-if-an-invalid-value-is-supplied-for-a-filter. Is the intent to change this in a future PR? Or is a decision being made to not track the errors?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py (hunk 7 above)
Comment (lzchen) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py:549]: Yes the purpose of this test was more to just check that `_check_custom_dim_field_filter` was called, but I changed the dimension to something meaningful regardless.

**Comment by lzchen** [comment]
@harsimar 

> When reading the config you may have converted things to lowercase for ease in filtering but I didn't see it (feel free to correct if I missed it!).

Yes this is indeed the case! There are `name.lower()` calls in both `_check_filters()` and `_validate_filter_field_name()`.

> Also note that exception related columns appear in the response as "Exception.Message" and "Exception.StackTrace".

There is a call to `_rename_exception_fields_for_filtering()` in `_update_filter_configuration()` to remove the `Exception.` part similar to how NodeJs implemented.

> Is the intent to change this in a future PR? Or is a decision being made to not track the errors?

Currently the intent is to not track the errors. I believe logging live metrics error would be extremely noisy (and Python customers tend to track ALL logs sent from their applications, including the ones from the SDK).

**Review by harsimar — APPROVED** [review]
Changes look good so far. Regarding tracking of config errors, I agree it's not a top priority thing to do at the moment. My understanding is that the tracking of the errors is mainly visible to the service side; it wouldn't be logged to console but sent to the post request body instead.

File: sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_types.py [hunk 8]
```diff
@@ -92,50 +91,52 @@ def _from_span(span: ReadableSpan):
...
-                span.attributes,
-            )
-        elif span.kind is SpanKind.PRODUCER:
-            dependency_type = "Queue Message"
-            msg_system = span.attributes.get(SpanAttributes.MESSAGING_SYSTEM)
-            if msg_system:
-                dependency_type += " | {}".format(msg_system)
-        else:
-            dependency_type = "InProc"
+        if span.attributes:
```
Comment (jeremydvoss) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_types.py:97]: Is this just an extra check or a special use case?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py [hunk 9]
```diff
@@ -214,9 +191,10 @@ def _update_filter_configuration(etag: str, config_bytes: bytes):
         # Skip duplicate ids
         if metric_info.id in seen_ids:
             continue
+        if not _validate_derived_metric_info(metric_info):
```
Comment (jeremydvoss) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py:194]: Could you explain this logic?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py [hunk 10]
```diff
@@ -246,6 +253,30 @@ def test_update_filter_configuration(
...
+    @mock.patch("azure.monitor.opentelemetry.exporter._quickpulse._utils._set_quickpulse_derived_metric_infos")
+    @mock.patch("azure.monitor.opentelemetry.exporter._quickpulse._utils._init_derived_metric_projection")
+    @mock.patch("azure.monitor.opentelemetry.exporter._quickpulse._utils._validate_derived_metric_info")
+    @mock.patch("azure.monitor.opentelemetry.exporter._quickpulse._utils.DerivedMetricInfo")
+    @mock.patch("azure.monitor.opentelemetry.exporter._quickpulse._utils._clear_quickpulse_projection_map")
+    def test_update_filter_configuration_invalid(
+        self, clear_mock, dict_mock, validate_mock, init_projection_mock, set_metric_info_mock, etag_mock
+    ):
+        etag = "new-etag"
+        test_config_bytes = '{"Metrics":[{"Id":"94.e4b85108","TelemetryType":"Request","FilterGroups":[{"Filters":[]}],"Projection":"Count()","Aggregation":"Sum","BackEndAggregation":"Sum"}]}'.encode()
```
Comment (jeremydvoss) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py:266]: What's invalid here?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py [hunk 11]
```diff
@@ -258,15 +242,216 @@ def _check_metric_filters(metric_infos: List[DerivedMetricInfo], data: _Telemetr
...
+
+def _validate_derived_metric_info(metric_info: DerivedMetricInfo) -> bool:
+    # Validate telemetry type
+    try:
+        telemetry_type = TelemetryType(metric_info.telemetry_type)
+    except Exception:  # pylint: disable=broad-except,invalid-name
+        return False
+    # Only REQUEST, DEPENDENCY, EXCEPTION, TRACE are supported
+    # No filtering options in UX for PERFORMANCE_COUNTERS
+    if telemetry_type not in (
```
Comment (jeremydvoss) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py:343]: What is this check for?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py [hunk 12]
```diff
@@ -258,15 +242,216 @@ def _check_metric_filters(metric_infos: List[DerivedMetricInfo], data: _Telemetr
...
-# pylint: disable=unused-argument
+# pylint: disable=R0911
 def _check_filters(filters: List[FilterInfo], data: _TelemetryData) -> bool:
+    if not filters:
+        return True
     # # All of the filters need to match for this to return true (and operation).
-    # for filter in filters:
-    #     # TODO: apply filter logic
-    #     pass
+    for filter in filters:
```
Comment (jeremydvoss) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py:250]: Could you summarize this logic?

File: sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_types.py (hunk 8 above)
Comment (jeremydvoss) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_types.py:97]: Safety check

File: sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py [hunk 13]
```diff
@@ -214,9 +191,10 @@ def _update_filter_configuration(etag: str, config_bytes: bytes):
         # Skip duplicate ids
         if metric_info.id in seen_ids:
             continue
+        if not _validate_derived_metric_info(metric_info):
+            continue
+        _rename_exception_fields_for_filtering(metric_info)
```
Comment (jeremydvoss) [inline sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py:196]: What's this?

Comment (jeremydvoss) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py:196]: To simplify handling the respond later. Remove "Exception." prefix.

File: sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py (hunk 9 above)
Comment (jeremydvoss) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py:194]: Make sure response makes sense and is in scope with what we support.

File: sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py (hunk 12 above)
Comment (jeremydvoss) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/azure/monitor/opentelemetry/exporter/_quickpulse/_utils.py:250]: This is not validation. This is the actual filtering logic.

File: sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py (hunk 10 above)
Comment (jeremydvoss) [reply sdk/monitor/azure-monitor-opentelemetry-exporter/tests/quickpulse/test_utils.py:266]: Validation is mocked to fail:
validate_mock.return_value = False
